### ISSUE TITLE
False positive reportAssignmentType for type widening from a newly-created object

### ISSUE THREAD

**Issue by Phrogz** [microsoft/pyright#10943]
**Describe the bug**

Code sample in [pyright playground](https://pyright-play.net/?code=GYexC4AIBsEsGcAuBtJAnAupAvJZAiAQ3wBpJ8AjfDAKAsLSjiVUTUgB9JhoRDEsuUCEgBqPPgDG1IA)

```python
foo: list[str] = ["a", "b"]
bar: list[str | float] = foo + ["c"]
```

This flags the assignment as invalid in the same way it would (correctly) if just `foo` was assigned. However, `foo + ["c"]` creates a new list with no other references to it, and so type widening should be accepted, as it is with just `bar: list[str | float] = ["a", "b", "c"]`

**VS Code extension or command-line**
First experienced in Cursor with `anysphere.cursorpyright` extension, but also present in `basedpyright` and (per above) reproduced directly in the pyright playground.

**Comment by LeeeeT** [comment]
Wow, this is interesting! I like it. I'll note that mypy has the same problem. You're hitting [this overload](https://github.com/python/typeshed/blob/00790cb021f8a2ea9306668636935a96d633dcca/stdlib/builtins.pyi#L1142). Both pyright and mypy infer `_S` as `str`, but it should be inferred as `str | float` instead. Here's a self-contained sample:

```python
from __future__ import annotations


class A[T]:
    v: T

    def __init__(self, v: T): ...
    
    def f[S](self, v: S) -> A[T | S]: ...


foo: A[str] = A("a")
bar: A[str | int] = foo.f("b")
```

**Comment by erictraut** [comment]
I think pyright is working correctly here. Both the LHS and RHS operands have the type `list[str]`. This means (contrary to what @LeeeeT said) that the first overload of `list.__add__` applies. The evaluated return result has the type `list[str]`.

As @LeeeeT mentioned, mypy generates an error here as well.

You can make this type check by changing your code to the following:

```python
bar = foo + list[str | float]([""])
```

**Comment by LeeeeT** [comment]
> the first overload of list.__add__ applies

Oh, yes. You're right!
